File: LeetCode/Study Plan/Practice Problems/Easy/Greatest Common Divisor Of An Array/Go/greatest_common_divisor_of_array.go
package main

import "fmt"

func greatest_common_divisor_of_array(arr []int) int {
	min, max := arr[0], arr[0]
	n := len(arr)
	for i := 0; i < n; i++ {
		if arr[i] < min {
			min = arr[i]
		}
		if arr[i] > max {
			max = arr[i]
		}
	}
	return gcd(min, max)
}

func gcd(a, b int) int {
	if b == 0 {
		return a
	}
	return gcd(b, a%b)
}

func main() {
	fmt.Println(greatest_common_divisor_of_array([]int{2, 5, 6, 9, 10}))
}

File: LeetCode/Easy/Greatest Common Divisor Of An Array/Cpp/greatest_common_divisor_of_array.cc
#include <iostream>
#include <numeric>
#include <vector>
#include <algorithm>

typedef unsigned long ui;
typedef std::vector<int> vi;

int greatest_common_divisor_of_array(vi &v) {
    size_t min_val=v[0], max_val=v[0];
    for (int i=0;i<v.size();i++) {
        min_val = std::min((ui)min_val, (ui)v[i]);
        max_val = std::max((ui)max_val, (ui)v[i]);
    }
    return std::gcd(min_val, max_val);
}

int main() {
    vi v{2,5,6,9,10};
    std::cout<<greatest_common_divisor_of_array(v)<<"\n";
}
